akka-stream/src/main/scala/akka/stream/impl/io/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ ActorLogging, DeadLetterSuppression, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.annotation.tailrec
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
```

Comment:
@ktoso I see it the other way round :) If you need guaranteed chunk sizes you can add an extra stage providing those. You cannot, however, revert the effect of holding data back by adding another stage. 

Actually, I find it strange (a code smell) that is even possible to hold data back. Reading data ahead to determine the end of the file makes only sense if you cannot know beforehand how much data is in the file. In such a case like when reading from a pipe (or if the file-size changes during reading), reading ahead won't help because it will close spontaneously. And, if you know the file size in advance you don't have to read ahead.